### ISSUE TITLE
Google Authentication, ChallengeResult, passing redirect uri correctly?

### ISSUE THREAD

**Issue by toddtsic** [aspnet/Mvc#2914]
I'm new to external login routines, but this is what has me stopped.  Would greatly appreciate assistance as I wire up Visual Studio 2015 mvc web app.

First attempt was with google.

Successfully set up google:

---------from google developer page----------------
Client ID: (xxed out here)
Client secret	(xxedout here)

Redirect URIs	
https://localhost:44300/tsic/Account/ExternalLoginCallback/
https://localhost/tsic/Account/ExternalLoginCallback/

JavaScript origins	
http://localhost:54618/
https://localhost:44300/
---------end from google developer page----------------

from Account controller:

// POST: /Account/ExternalLogin
    [HttpPost]
    [AllowAnonymous]
    [ValidateAntiForgeryToken]
    public IActionResult ExternalLogin(string provider, string returnUrl = null)
    {

        // Request a redirect to the external login provider.
        var redirectUrl = String.Format(
            "{0}://{1}{2}/", 
            "https", 
            Request.Host.ToString(),
            Url.Action("ExternalLoginCallback", "Account", new { jSeg = "tsic"})
            //Url.Action("ExternalLoginCallback", "Account", new { jSeg = cd.JobParams.path, ReturnUrl = returnUrl })
        );
        var properties = SignInManager.ConfigureExternalAuthenticationProperties(provider, redirectUrl);
        return new ChallengeResult(provider, properties);
    }

I then compared the returnUrl from the controller with that from google's developers page:

from google:                         https://localhost:44300/tsic/Account/ExternalLoginCallback/
from controller in debugger: https://localhost:44300/tsic/Account/ExternalLoginCallback/

confirming that they are identical.

Then executing the code, I do get to login to google but get immediately (from the network tab of chrome inspect elements):

400. That’s an error.

    Error: redirect_uri_mismatch

    Application: TeamSportsInfo.com

    You can email the developer of this application at: ToddTSIC@gmail.com

    The redirect URI in the request: https://localhost:44300/signin-google did not match a registered redirect URI.

    Learn more

    Request Details
    scope=openid profile email
    response_type=code
    redirect_uri=https://localhost:44300/signin-google
    state=CfDJ8OmOKsR1RZdPli5yiRwe3RPk8sA5U-OBSLSchc8NqzN5r5EIPXc9-C89BEm8SdW6n1XoF3xIvwlV-s4X8vWblMihbAs5_Kmmii32PhvqnCX4eDSeOgnXALohBmJ1pddngXalP_jZket5iTQ4S5aJM5uzra9IHgENcGJSuvlgGtnB9Vhgwz5rWdh_nWZobZpIPYofTIlqRvtueUCVcTM5LDZg22r0EMloT68B8hMdJ1uv7xLq36U-TrSaEz4R7hLUCx7ouSofBnETNNM5CBFBs7l4GAPdaTTn72ojP-oDjZiEqJQ_8g11NKubpyViPfL4V68jXI2RU41wLKWXcGNmgI4
    client_id=368633669455-5tlpc2c3jgh8r6u7l76ldktdpasd4brc.apps.googleusercontent.com
    That’s all we know.

What strikes me is the line from the google error:

redirect_uri=https://localhost:44300/signin-google

this after setting the redirectUrl of SignInManager.ConfigureExternalAuthenticationProperties(provider, redirectUrl);

to: https://localhost:44300/tsic/Account/ExternalLoginCallback/

It makes sense to me that google is protesting at redirect_uri=https://localhost:44300/signin-google
as that is not what my google Redirect URIs are set to.

This suggests to me that 1) I'm missing something obvious, or 2) var properties = SignInManager.ConfigureExternalAuthenticationProperties(provider, redirectUrl);
            return new ChallengeResult(provider, properties); is not passing the indicated returnUrl

**Comment by toddtsic** [comment]
Addendum:

Looking at the request headers sent to google confirms that my code is sending an incorrect querystring "redirect_uri":

From chrome network tools:

Remote Address:216.58.218.205:443
Request URL:https://accounts.google.com/o/oauth2/auth

?response_type=code

&client_id=xxdout

&redirect_uri=https%3A%2F%2Flocalhost%3A44300%2Fsignin-google (THIS IS INCORRECT AND IS NOT WHAT WAS SUPPLIED BY SignInManager.ConfigureExternalAuthenticationProperties)

&scope=openid%20profile%20email

&state=xxdout

**Comment by toddtsic** [comment]
Just repeated process with Facebook and the exact same problem, the redirect_uri sent to facebook is not what was configured in ConfigureExternalIAuthenticationPropertiers:

https://www.facebook.com/v2.2/dialog/oauth
?client_id=xxdout
&scope=
&response_type=code
&redirect_uri=https%3A%2F%2Flocalhost%3A44300%2Fsignin-facebook (THIS IS NOT WHAT WAS CONFIGURED)
&state=xxdout

**Comment by Eilon** [comment]
Indeed, @HaoK can you take a look?

**Comment by HaoK** [comment]
So the issue is that you are configuring the final redirect (not the google/fb redirects) when configuring sign in manager's external properties.  To configure google/fb's redirects, you need to configure them via their options:

CallbackPath is the uri that google/fb will callback into.

https://github.com/aspnet/Security/blob/dev/src/Microsoft.AspNet.Authentication.OAuth/OAuthAuthenticationOptions.cs

**Comment by toddtsic** [comment]
I have figured this out.  Was a misunderstanding on my part, but I believe my error will provide some insight on how to make clearer for others:

The key to unraveling this was that 
    with google the mvc produced redirect_uri was hostname://signin-facebook
   and for facebook: hostname://signin-facebook

From this I concluded that the redirect uri to be set at google and facebook developer setup pages should be those values and not (and here's my misunderstanding), the location of the callback method in mvc code: ExternalLoginCallback

So setup instructions should be:

1) set your provider's (Google, Facebook), redirect_uri to be https://hostnamexxxx/signin-(providername)  (there was no way to know this value other than by tracing the headers produced by ChallengeResult(provider, properties) in Account/ExternalLogin)

2) set the redirectUrl for SignInManager.ConfigureExternalAuthenticationProperties(provider, redirectUrl) to be your location of ExternalLoginCallback method.

To restate my mistake: I thought the redirect_uri set on the provider setup side (at Google or at Facebook) was to be the url of the mvc callback method: ExternalLoginCallback.  It's not, rather it must be protocol://hostname/signin-(providername).  The url of the callback method: ExternalLoginCallback however must be the redirectUrl to be used in ExternalLogin method's call to ConfigureExternalAuthenticationProperties.
